### ISSUE TITLE
Support 'auto' dark theme that reflects OS settings

### ISSUE THREAD

**Issue by Jaifroid** [kiwix/kiwix-js#729]
There are now CSS media queries and corresponding DOM functions in Firefox and Chromium (and Safari) that can detect the system-wide / OS 'dark mode' settings in both Windows and Mac Mojave (I haven't checked for Linux yet). See: https://caniuse.com/?search=prefers-color-scheme.

I've tested this in Kiwix JS Windows, and simple code like this can register to be notified of a change in the system preferences, and adapt the auto theme accordingly:
```
var uiSettings = window.matchMedia('(prefers-color-scheme:dark)');
uiSettings.onchange = function () {
    if (params.cssUITheme == 'auto') cssUIThemeGetOrSet('auto');
}; 
```
The bit inside the `onchange` function would need to be adapted to Kiwix JS code, but we have the methods for switching themes. Some OS's allow auto switching themes according to time of day, for example, so Kiwix would auto-adapt when the system changes the theme, or simply when the user switches the OS theme manually. I've tested this on Windows 10, with the global Light/Dark theme in Settings.

Theme should of course default to Light if the user is using an older browser where the OS theme cannot be detected. The user can override by choosing Light or Dark explicitly, instead of, say, Auto, in Kiwix JS Configuration.

**Comment by ShubhashishMaiti** [comment]
Hello Sir, I am interested in working on this issue. Could you please assign this issue to me?

**Comment by Jaifroid** [comment]
@ShubhashishMaiti That would be helpful. Please note that the sample code I gave above was for a related app, Kiwix JS Windows, where this is implemented. It would need to be adapted to the parameters used in Kiwix JS, but the principle is the same.

**Comment by Maneet-k** [comment]
Sir, I wanna work on this issue. Can I help you in this ?

**Comment by Maneet-k** [comment]
@mossroy are nightwatch.js and nightwatch_runner.js  ,the files I should look into?

**Comment by mossroy** [comment]
No, these files are used for automated testing, and have nothing to do with a night theme

**Comment by Maneet-k** [comment]
@mossroy ohh my bad....... would you please mind me telling the files i should look into. That'll be really helpful🤧

**Comment by Jaifroid** [comment]
@Maneet-k  The app themes are instantiated with the parameter `appTheme`. If you search for `params.appTheme` in `app.js`, you should be able to figure out how it works.. Also look at the utility function `applyAppTheme` in  `uiUtil.js`. You don't need to change any CSS or add any other JS files. You just need to add a function in `app.js`, that listens on a system theme change and applies it using the function in `uiUtil.js`. You can find sample code in my first comment above (obviously it has to be adapted to kiwix.js).

**Comment by Jaifroid** [comment]
You'll also need to add an 'auto' option to the existing theme UI, and decide which of the dark themes will be applied with the auto option. The most useful one for users would probably be the Wikimedia dark theme, but that might need some debate amongst the devs.

**Comment by Maneet-k** [comment]
@Jaifroid I added the function in the app.js as guided by you. But I'm not able to understand the later one change. Is it if the user wants to override their system preferences theme. Then we need a button right? So they could toggle b/w the themes?
Is that what you meant in the issue?

**Comment by Jaifroid** [comment]
@Maneet-k You don't need an extra button. You just need to add an option "Auto (uses system theme)" to the drop-down list of available themes. I suggest adding "Auto (uses system theme)" as the top option, but that "Light" should still be the default.

When the user selects "Auto", the system should choose between "Light" and "Dark + Invert Wikimedia" according to whether the system theme is Light or Dark.

However, @mossroy should decide whether he would prefer the auto dark theme to be the generic, non-Wikimedia dark theme (the one called "Dark + invert generic"). I can see pros and cons.

**Comment by mossroy** [comment]
I agree there are pros and cons.
Maybe we should offer both options : 

- auto (generic)
- auto (Wikimiedia)*

(with the same * that displays the same warning)

If necessary, this can be done afterwards

**Comment by Jaifroid** [comment]
OK, then I'd suggest starting with "Auto (generic)" precisely because it is generic!

**Comment by Maneet-k** [comment]
@Jaifroid is should I make a prompt for two these two dark theme for user if his system preference is dark. Or I should just go with the generic one.

**Comment by Jaifroid** [comment]
I don't think a "prompt" is needed (if you mean something like a dialogue box asking which one the user wants). I suggest you just code the generic dark theme for now. Once it is done, it would be easy to add the second one if necessary.

**Comment by Maneet-k** [comment]
also Also, what exact5ly is "auto".   is it a value  you've given to appTheme or this is just for the reference to implement the things? @Jaifroid

**Comment by Jaifroid** [comment]
There is no existing "auto" value. It is just what we are calling the mode. You will need to have an entry in the dropdown menu that allows the user to select the theme. It should read "Auto (generic)", and be at the top, but the default should still be "Light".

**Comment by Maneet-k** [comment]
<img width="668" alt="2021-12-12 (2)" src="https://user-images.githubusercontent.com/81208462/145719526-7d03a5c8-1ea4-43fc-9ac8-785be235eeae.png">
@Jaifroid Could please check if this will work?

**Comment by Maneet-k** [comment]
option1 : Auto(Generic) [value: dark_invert]
option2 : Auto(Wikimedia) [value: dark_winvert]
option3 : Light [value: light]
option4 : dark [value: dark]
are as values in the the dropdown

**Comment by Jaifroid** [comment]
There are currently two manual Dark options (generic and wikimedia). You should keep those.

**Comment by Jaifroid** [comment]
> @Jaifroid Could you please check if this will work?

You should try it out and see what works. It's hard to tell in the abstract...

**Comment by Maneet-k** [comment]
@Jaifroid sorry, I'm really bothering you. But don't you think if user wants to override it's preference for this particular app from dark to light or vice versa. We should provide them those options too. Instead of "auto" only. wouldn't that be better. Like if we just provide the light and dark(generic or wikimedia) .
Flow of the execution:
first it will automatically decide the system theme as according to it's system preference. But then if user wanted to change the theme. He/She can do this manually.
It's just my suggestion to make app work better.
You see whatever fits in right. I'll be working on that one .

**Comment by Jaifroid** [comment]
@Maneet-k Of course, we need to keep the manual options. When the user sets the 'Auto' option, you should set a global parameter like `params.autoDarkMode = true;`. This would then enable the event listener which would control the setting or unsetting of the dark mode. When the user sets one of the manual options, this event listener should either be disabled or simply blocked (with a test on `params.autoDarkMode`), so that it doesn't interfere with the manual setting.

**Comment by Jaifroid** [comment]
It may be possible to do it without a global parameter, if the Auto setting enables the event listener, and a manual setting disables the event listener.

**Comment by Maneet-k** [comment]
@Jaifroid Thanks a lot sir for your assistance and precious time💖, I really appreciate. But I'm still not able to understand the code. I'll do check the solution when posted by you or someone else. Please unassign me from this task. My apologize.
Have a Merry Christmas ✨🦄

**Comment by Jaifroid** [comment]
OK, no problem. Thanks for trying.

**Comment by bellaryyash23** [comment]
Hello @Jaifroid Sir, 
I have read the above discussions and issue and I would like to provide my inputs. So I request you to please assign me this issue.
Thank you

**Comment by Maneet-k** [comment]
Things where i got confused when the two auto dark themes came into the convo. I wasn't able to figure out if i need to make a drop-down box /toggle for them or not.

**Comment by Jaifroid** [comment]
@bellaryyash23 Thanks for the offer. I'm assigning you. Please experiment and test your ideas/code carefully. You will need to serve your code to your local browser with something like `http-server` (node/npm), and access the code from http://localhost:8080/ . We test on Chromium (Chrome/Edge) and Firefox as a minimum, ideally also IE11 or IE mode in Edge.

**Comment by Maneet-k** [comment]
Hi @bellaryyash23 , How you doin?
Actually ,I resolved this issue. Can I raise PR? If you don't mind.

**Comment by Jaifroid** [comment]
Yes, anyone can make a PR, but please be sure you've tested it thoroughly.

**Comment by bellaryyash23** [comment]
ok @Maneet-k . @Jaifroid can you please check the PR. If any updates or tweaks are there then I would be ready to help.

**Comment by bellaryyash23** [comment]
Sorry I was unable to solve the issue. But it was a great experience. Thank you @Jaifroid. You can now unassigned me.

**Comment by Jaifroid** [comment]
@Maneet-k Do you still want to work on this?

**Comment by Maneet-k** [comment]
@Jaifroid , I got issues while testing on different cases. 
Ok I added the auto option with auto value
But this auto value is not getting changed to dark_invert or light if accessed via dom. or simply applying the dark or light using uiUtil.applyAppTheme("dark_invert") if darkPreferences.matches.
But I still want to work on this

**Comment by Maneet-k** [comment]
could you tell me where i am doin' wrong

**Comment by Jaifroid** [comment]
@Maneet-k If you make a pull request with your current code, I can comment on it and point out any obvious issues. Without seeing actual code, it's hard to know where you might be going wrong.

**Comment by Maneet-k** [comment]
in index.html:
<option value = "auto"> Auto<option>

**Comment by Maneet-k** [comment]
@Jaifroid  Could you please test the changes that I've commited

**Comment by Maneet-k** [comment]
@Jaifroid sir, How can I make this work across cross browsers. It's working well on edge, Chrome and firefox. But still github pages throwing err while testing for Cross-browser.

**Comment by Jaifroid** [comment]
@Maneet-k I'm not sure how you're testing. Which GitHub pages implementation are you accessing?

I'm not seeing specific test failures. It's OK to test manually only. It might be easiest for you to serve from your own localhost, using something like 'http-server' from node/npm.

Please also see [this comment](https://github.com/kiwix/kiwix-js/issues/790#issuecomment-1008669638) about needing to empty the appCache after you have changed code and before reloading the app to see the new changes. This will be the case for any branch you have rebased on master (like #793).

**Comment by Maneet-k** [comment]
@Jaifroid I don't know why but this http-server method has stopped working for me. Whenever I try to serve the app using this method , terminal gets crashed. But I got the other way to serve the app. Do I still need to follow the process of deleting cache?

**Comment by Maneet-k** [comment]
It's working today. That must be a problem in my system other day

**Comment by Jaifroid** [comment]
> Do I still need to follow the process of deleting cache?

You will need to do this if you find that the code doesn't update in your browser after you make changes. Personally I delete the cache before reloading routinely at the moment, to be sure.

**Comment by Maneet-k** [comment]
Ohkayy, @Jaifroid

**Comment by aritroCoder** [comment]
Is the issue still open? Can I work on this?

**Comment by Jaifroid** [comment]
@aritroCoder There is an open PR already for this, which is in the process of being tested (I think), so you should probably choose another issue.
